frontends/concrete-python/concrete/fhe/compilation/module.py

```diff
@@ -0,0 +1,669 @@
+    def encrypted_negation_count_per_tag(self) -> Dict[str, int]:
+        """
+        Get the number of encrypted negations per tag in the function.
+        """
+        return self.runtime.server.encrypted_negation_count_per_tag(self.name)  # pragma: no cover
+
+    @property
+    def encrypted_negation_count_per_tag_per_parameter(self) -> Dict[str, Dict[Parameter, int]]:
+        """
+        Get the number of encrypted negations per tag per parameter in the function.
+        """
+        return self.runtime.server.encrypted_negation_count_per_tag_per_parameter(
+            self.name
+        )  # pragma: no cover
+
+    @property
+    def statistics(self) -> Dict:
+        """
+        Get all statistics of the function.
+        """
+        attributes = [
+            "size_of_inputs",
+            "size_of_outputs",
+            "programmable_bootstrap_count",
+            "programmable_bootstrap_count_per_parameter",
+            "programmable_bootstrap_count_per_tag",
+            "programmable_bootstrap_count_per_tag_per_parameter",
+            "key_switch_count",
+            "key_switch_count_per_parameter",
+            "key_switch_count_per_tag",
+            "key_switch_count_per_tag_per_parameter",
+            "packing_key_switch_count",
+            "packing_key_switch_count_per_parameter",
+            "packing_key_switch_count_per_tag",
+            "packing_key_switch_count_per_tag_per_parameter",
+            "clear_addition_count",
+            "clear_addition_count_per_parameter",
+            "clear_addition_count_per_tag",
+            "clear_addition_count_per_tag_per_parameter",
+            "encrypted_addition_count",
+            "encrypted_addition_count_per_parameter",
+            "encrypted_addition_count_per_tag",
+            "encrypted_addition_count_per_tag_per_parameter",
+            "clear_multiplication_count",
+            "clear_multiplication_count_per_parameter",
+            "clear_multiplication_count_per_tag",
+            "clear_multiplication_count_per_tag_per_parameter",
+            "encrypted_negation_count",
+            "encrypted_negation_count_per_parameter",
+            "encrypted_negation_count_per_tag",
+            "encrypted_negation_count_per_tag_per_parameter",
+        ]
+        return {attribute: getattr(self, attribute) for attribute in attributes}
+
+
+class FheModule:
+    """
+    Fhe nodule class, to combine computation graphs, mlir, runtime objects into a single object.
+    """
+
+    configuration: Configuration
+    graphs: Dict[str, Graph]
+    mlir_module: MlirModule
+    compilation_context: CompilationContext
+    runtime: Union[ExecutionRt, SimulationRt]
+
+    def __init__(
+        self,
+        graphs: Dict[str, Graph],
+        mlir: MlirModule,
+        compilation_context: CompilationContext,
+        configuration: Optional[Configuration] = None,
+    ):
+        assert configuration and (configuration.fhe_simulation or configuration.fhe_execution)
```

Comment:
This shouldn't be an assetion, we can allow MLIR generation without compilation, which makes some sense. Simulation and Execution can be enabled later on demand anyway.

Reply:
It is not what I did for the multi-function. The runtime is not lazilly instantiated, but rather is instantiated upfront. I can change it, but it requires a lot of changes.

Reply:
It's not that important but it'd be nice to support in the future maybe. Good to go!